### PR TITLE
Use MC Stats for PHP fatal error counting

### PR DESCRIPTION
**Pull request by chihsuan**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

Part of https://github.com/woocommerce/woocommerce/issues/49289

This PR creates a McStats class which extends from [Jetpack's `A8c_Mc_Stats`](https://github.com/Automattic/jetpack-a8c-mc-stats/blob/trunk/tests/php/test_Stats.php) class for recording stats, and use it to record fatal errors.

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Create a fresh site using this branch with code snippet plugin installed and activated.
2. Go to Core Profiler and allow tracking
3. Add the following code snippet to the code snippet plugin:

```php
function custom_wp_remote_get_debug( $response, $args, $url ) {
    if ( strpos( $url, 'https://pixel.wp.com' ) === 0 ) {
        // Print the request URL
        echo 'Request URL: ' . esc_url( $url ) . '<br>';

        // Check if the response is a WP_Error or not
        if ( is_wp_error( $response ) ) {
            echo 'Response: Failed<br>';
        } else {
            echo 'Response: Success<br>';
        }
    }
    return $response;
}

add_filter( 'http_response', 'custom_wp_remote_get_debug', 10, 3);
```

4. Trigger a fatal error in WooCommerce by adding the following code to [load_rest_api()](https://github.com/woocommerce/woocommerce/blob/ac834f55f75d4e51955bad43115939faa38ec9da/plugins/woocommerce/includes/class-woocommerce.php#L535):

```php
throw new Exception( 'Test fatal error in woocommerce' );
```
5. Reload the page
6. Confirm that the screen shows the `Response: Success` 

Screenshot 2024-07-24 at 10 59 17

### Changelog entry

-   [ ] Automatically create a changelog entry from the details below.

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message 

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>

### CODE REVIEW

File: plugins/woocommerce/includes/class-woocommerce.php [hunk 1]
```diff
@@ -403,6 +404,12 @@ public function log_errors() {
 				$context
 			);
 
+			// Record fatal error stats.
```
Comment (ilyasfoo) [inline plugins/woocommerce/includes/class-woocommerce.php:406]: I think we need to respect user's consent here? Since this is calling an external API 🤔

**Review by ilyasfoo — COMMENTED** [review]
I confirmed it triggered a call to pixel, but I too wasn't able to see the stats after 30 minutes.

image

I think we need to not record the stats when we don't have user's consent. What do you think?

File: plugins/woocommerce/includes/class-woocommerce.php (hunk 1 above)
Comment (chihsuan) [reply plugins/woocommerce/includes/class-woocommerce.php:406]: You're right. I had that in the issue acceptance criteria but forgot to implement it. 😓  I'll update the code.

File: plugins/woocommerce/src/Stats/McStats.php [hunk 2]
```diff
@@ -0,0 +1,35 @@
+<?php
+declare( strict_types = 1 );
+
+/**
+ * WooCommerce MC Stats package
+ */
+
+namespace Automattic\WooCommerce\Stats;
+
+use Automattic\Jetpack\A8c_Mc_Stats;
```
Comment (vedanshujain) [inline plugins/woocommerce/src/Stats/McStats.php]: Does this work if Jetpack is not available on a site?

Comment (ilyasfoo) [reply plugins/woocommerce/src/Stats/McStats.php]: Good question! This is using the [imported class](https://github.com/Automattic/jetpack-a8c-mc-stats/blob/6b7a74bd7c08773d3a156877cad20290ac87f7e4/src/class-a8c-mc-stats.php#L13) which is added in composer dependency [here](https://github.com/woocommerce/woocommerce/pull/49658/files#diff-62b5ff93172c7be053df8541c8702199d88e52766e3e7f3dc215dd5b6dfe22d4R22). Thus, this shouldn't rely on Jetpack plugin.

Comment (chihsuan) [reply plugins/woocommerce/src/Stats/McStats.php]: Yes, like @ilyasfoo said, this should work without Jetpack plugin.

Comment (moon0326) [reply plugins/woocommerce/src/Stats/McStats.php]: Are we expecting any other variations of stat classes? If not, I feel like we can just put the file in the `Internal` directory

File: plugins/woocommerce/includes/class-woocommerce.php (hunk 1 above)
Comment (moon0326) [reply plugins/woocommerce/includes/class-woocommerce.php:406]: It seems `McStats` is going to be used exclusively in `class-woocommerce.php` and does not require any constructor dependencies. I'm not against it, but I think we don't need to add this class to the container when it's not meant to be used any other places.

**Review by vedanshujain — COMMENTED** [review]
Apologies for another drive-by comment, but curious as to why are we sending this from frontend? Seems like this is a sensitive info that the site owner may not want to expose.

File: plugins/woocommerce/includes/class-woocommerce.php (hunk 1 above)
Comment (chihsuan) [reply plugins/woocommerce/includes/class-woocommerce.php:406]: I believe we will use McStats in other places in the future. It's better than using tracks to count events. Does it hurt to add it to the container now? 🙂

File: plugins/woocommerce/src/Stats/McStats.php (hunk 2 above)
Comment (chihsuan) [reply plugins/woocommerce/src/Stats/McStats.php]: Not that I know of for now. I think that's a good idea. I'll move it there.

**Comment by chihsuan** [comment]
> Apologies for another drive-by comment, but curious as to why are we sending this from frontend? Seems like this is a sensitive info that the site owner may not want to expose.

Hey @vedanshujain  No worries! It's just because it's easier to test, and I thought it would reduce the load on the server (though it's very minimal). 

Could you share more about what sensitive info you're referring to?  I'm okay with sending it from the server side.

File: plugins/woocommerce/src/Stats/McStats.php (hunk 2 above)
Comment (chihsuan) [reply plugins/woocommerce/src/Stats/McStats.php]: Moved `McStats` to `./src/Internal` in d726f0c7ccad976d381c344e623829b8a7416edc

File: plugins/woocommerce/includes/class-woocommerce.php (hunk 1 above)
Comment (chihsuan) [reply plugins/woocommerce/includes/class-woocommerce.php:406]: Done, I added a check for user's consent before recording the stats in  fe088e72d572bbf840b60601a1c890d6ba3508bc

Comment (chihsuan) [reply plugins/woocommerce/includes/class-woocommerce.php:406]: @ilyasfoo I'm seeing the stats now. 😃 

1. Go to MC Stats
2. Click on `Simple List - active`
3. Search for `woocommerce-error`

**Comment by chihsuan** [comment]
I've changed it to send from the backend.  Also updated the test instructions to reflect this change.  bed261abb93802097e5d58f70f90f5858c69512a

File: plugins/woocommerce/includes/class-woocommerce.php (hunk 1 above)
Comment (moon0326) [reply plugins/woocommerce/includes/class-woocommerce.php:406]: @chihsuan In that case, nope :) It makes perfect sense.

Comment (ilyasfoo) [reply plugins/woocommerce/includes/class-woocommerce.php:406]: >Done, I added a check for user's consent before recording the stats in https://github.com/woocommerce/woocommerce/commit/fe088e72d572bbf840b60601a1c890d6ba3508bc

@chihsuan Nice, thanks! Consent check looks good

File: plugins/woocommerce/src/Internal/McStats.php [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+	/**
+	 * Runs stats code for a one-off, server-side.
+	 *
+	 * @param string $url string The URL to be pinged. Should include `x_jetpack-{$group}={$stats}` or whatever we want to store.
+	 *
+	 * @return bool If it worked.
+	 */
+	public function do_server_side_stat( $url ) {
+		if ( ! \WC_Site_Tracking::is_tracking_enabled() ) {
+			return;
```
Comment (ilyasfoo) [inline plugins/woocommerce/src/Internal/McStats.php]: I think we need to return bool here
```suggestion
			return false;
```

File: plugins/woocommerce/tests/php/src/Internal/McStatsTest.php [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+		$this->sut->add( 'group', 'test2' );
+
+		$this->assertEmpty( $this->sut->get_group_query_args( 'group2' ) );
+
+		$check = $this->sut->get_group_query_args( 'group' );
+
+		$this->assertCount( 1, $check );
+		$this->assertArrayHasKey( 'x_woocommerce-group', $check );
+		$this->assertEquals( 'test,test2', $check['x_woocommerce-group'] );
+	}
```
Comment (ilyasfoo) [inline plugins/woocommerce/tests/php/src/Internal/McStatsTest.php:35]: I think it'll be good to add a test for consent check here

**Review by ilyasfoo — COMMENTED** [review]
Testing well, thanks @chihsuan! Minor comments:

File: plugins/woocommerce/tests/php/src/Internal/McStatsTest.php (hunk 4 above)
Comment (chihsuan) [reply plugins/woocommerce/tests/php/src/Internal/McStatsTest.php:35]: That's a good idea.  Added here: 2aae131e54d0ba580432fbbb53eb007bbaac07e1

**Review by ilyasfoo — APPROVED** [review]
Thanks, LGTM! 🚢

**Review by moon0326 — APPROVED** [review]
LGTM! 👍 🚀

**Pull request merged**